### PR TITLE
feat(auth) - Phaseout mjml auth server

### PR DESCRIPTION
**Pull request by dschom**
## Because

- A separate server was required to render templates needed for storybook.

## This pull request

- Creates some bindings to abstract away process context used by FluentLocalizer.
- These bindings support running in either a nodejs process or in the browser.

## Issue that this pull request solves

Closes: #10218

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: packages/fxa-auth-server/package.json [hunk 1]
```diff
@@ -31,9 +31,8 @@
     "paypal-processor": "CONFIG_FILES='config/secrets.json' node -r esbuild-register ./scripts/paypal-processor.ts",
     "subscription-reminders": "CONFIG_FILES='config/secrets.json' node -r esbuild-register ./scripts/subscription-reminders.ts",
     "emails-scss": "node -r esbuild-register ./lib/senders/emails/sass-compile-files.ts",
-    "start-mjml": "NODE_ENV=dev node -r esbuild-register ./bin/mjml-server.ts",
-    "storybook": "npm run emails-scss && start-storybook -p 6010 --no-version-updates",
-    "storybook-mjml": "nodemon -e scss,ts  --exec \"grunt merge-ftl && npm-run-all -p watch-ftl start-mjml storybook\"",
+    "storybook": "npm run install-ejs && npm run emails-scss && start-storybook -p 6010 --no-version-updates -s ./",
```
Comment (LZoog) [inline packages/fxa-auth-server/package.json:34]: 🎉 I didn't like calling it `storybook-mjml` anyway even when we had the mjml server. 😝

**Review by LZoog — COMMENTED** [review]
@dschom, I didn't check into the branch since you still have this set as a WIP draft but I did a first pass review and I want to give you a big high five here. 😄 I think essentially having that separate config (and helper/tag replacer) for viewing MJML in the browser is a great approach. Thanks again for all of your effort helping us finish out this project, your perspective has been really appreciated.

File: packages/fxa-auth-server/lib/senders/emails/localizer-bindings-node.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import { readFileSync } from 'fs';
+import { JSDOM } from 'jsdom';
+import {
+  LocalizerBindings,
+  LocalizerOpts,
+  TemplateContext,
+} from './localizer-bindings';
+
+import ejs = require('ejs');
+import mjml2html = require('mjml');
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-auth-server/lib/senders/emails/localizer-bindings-node.ts:14]: Ah good catch! This is how it was imported previously, but you're right I don't think it's necessary.

File: packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  // <mj-style> tags must go in header. Create one if possible,
+  // otherwise error out.
+  const hasOpeningMjHeadTag = /<mj-head>/.test(mjml);
+  const hasClosingMjHeadTag = /<\/mj-head>/.test(mjml);
+  if (!hasOpeningMjHeadTag && !hasClosingMjHeadTag) {
+    mjml = mjml.replace(
+      '<mjml>',
+      `<mjml>
+  <mj-head>
+  </mj-head>`
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts]: We could have it error out, and I was very tempted to do so. If I recall correctly, the Node version of mjml would automatically handle this case, which is why I tried to support it.

I don't think the spacing matters. I'll shore this up. You're right, the code will look cleaner this way.

File: packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+  // Parse mjml and build style statements using ejs includes
+  const includes = extractMjIncludeTags(mjml);
+
+  // Append includes to end of header
+  if (includes.length) {
+    // Update the header tag, appending the includes
+    mjml = mjml.replace(
+      /\<\/mj-head\>/,
+      '\n  ' +
```
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts]: In my opinion, this operation should affect resulting mjml as little as possible. In my mind, there's no benefit to removing the formatting in the mjml just for parsing it and adding a few new lines. That being said, it probably doesn't matter if all the injected includes are on a single line. I can clean this up, so the code looks nicer. 

For the unit test, I found it easier to write tests and spot problems in test failures when the HTML was compacted, that's why that function you reference is in there. I could probably take it out, but in the event there was some stray white space, it made it almost impossible to see the issue.

On a side note, I was  tempted to use DOMParser to do this, but we are in this quasi state where the tests are running in node, but this code will execute in the browser. For this reason, I decide to keep it as vanilla JavaScript. Perhaps I should have found a way to inject a DOMParser...

**Comment by LZoog** [comment]
@dschom I'll take a closer at your PR shortly since you've pushed some changes, but if you click on "Details" in CI for `storybooks: pull request` and go to the auth-server's Storybook deployment, there's a "Loading error - Missing tag". Not sure if directly related since I wouldn't call this a "tag" but in the console, there's: `[fluent] Missing translations in en: subscriptionCancellation-subject`

File: packages/fxa-auth-server/README.md [hunk 5]
```diff
@@ -258,8 +258,13 @@ http://localhost:3030/connect_another_device?utm_medium=email&utm_campaign=fx-ca
...
+
 Each template has an optional `doc` string which serves to document when users will receive the email meaning what triggers the email to be sent. These templates are essentially HTML templates rendered in browser with Storybook and there is a possibility that they may not exactly match with what users see in an email client.
-To preview these templates with Storybook, run `yarn storybook-mjml` in `fxa-auth-server` root directory
+To preview these templates with Storybook, run `yarn storybook` in `fxa-auth-server` root directory.
+
+In addition to running locally, Storybook can be built into a static format. This is ultimately what happens in our CI. This static format is then deployed and manual QA may be conducted against it. To test what this will look like, run `yarn build-storybook` then navigate to the `storybook-static` folder and run `http-server . -p 8081`. From there, simply navigate to the `http://locahost:8081/index.html` and you will see the resulting static build. _(If you don't have `http-sever` installed, simply run `npm install -g http-server`)_
```
Comment (LZoog) [inline packages/fxa-auth-server/README.md:265]: Thank you for all the doc updates!! We could also potentially add the `fxa-auth-server` to the list under the [heading here](https://github.com/mozilla/fxa/#storybook-in-circleci).

File: packages/fxa-auth-server/lib/senders/emails/localizer-bindings-browser.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+// storybook starts up.
+import ejs from '../../../vendor/ejs';
+import { transformMjIncludeTags } from './mjml-browser-helper';
+
+/**
+ * Allows ejs to import requested files. This gets invoked when include() is
+ * used in a template. Note that this function MUST be synchronous.
+ */
+ejs.fileLoader = function (filePath: string) {
+  var request = new XMLHttpRequest();
```
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-auth-server/lib/senders/emails/localizer-bindings-browser.ts:28]: 👀

File: packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (LZoog) [reply packages/fxa-auth-server/lib/senders/emails/mjml-browser-helper.ts]: Cool yeah, after looking more closely after your cleanup it looks great. My main concern was that I thought we were explicitly replacing something like, `  <whatever` which was dependent on the number of spaces which isn't guaranteed.

Eh yeah, I think it's okay without. 🙂

**Review by LZoog — APPROVED** [review]
👏👏

This is such great work @dschom. Thanks again for tackling this problem. I had just one or two nitpicks left but I'm so stoked to see the deployment working properly and things work locally as well. When I run `yarn storybook` I see a deprecation warning but I think that's tied to `sass-compile-files` and is probably something I didn't notice before. 

:shipit:

File: packages/fxa-auth-server/README.md (hunk 5 above)
Comment (dschom) [reply packages/fxa-auth-server/README.md:265]: Link added.

**Comment by jodyheavener** [comment]
🎉

**Pull request merged**